lib/iris/tests/test_merge.py

```diff
@@ -145,81 +145,143 @@ def _make_cube(self, data, dtype=np.dtype('int32'), fill_value=None,
         cube.add_aux_coord(height)
         return cube
 
+    @staticmethod
+    def _expected_fill(fill0, fill1):
+        result = None
+        if fill0 == fill1:
+            result = fill0
+        return result
+
+    def _check_fill_value(self, result, fill0, fill1):
+        fill_value = self._expected_fill(fill0, fill1)
+        if fill_value is None:
+            fill_value = ma.masked_array(0, dtype=result.dtype).fill_value
+            self.assertIsNone(result.fill_value)
+            data = result.data
+            if ma.isMaskedArray(data):
+                self.assertEqual(data.fill_value, fill_value)
+        else:
+            self.assertEqual(result.fill_value, fill_value)
+            data = result.data
+            if ma.isMaskedArray(data):
+                self.assertEqual(data.fill_value, fill_value)
+
     def setUp(self):
         self.dtype = np.dtype('int32')
-        self.fill_value = 1234
-        self.lazy_combos = itertools.product([False, True],
-                                             [False, True])
+        fill_value = 1234
+        lazy_combos = itertools.product([False, True],
+                                        [False, True])
+        fill_combos = itertools.product([None, fill_value],
+                                        [fill_value, None])
+        self.combos = itertools.product(lazy_combos, fill_combos)
```

Comment:
Was it too complicated to have another `itertools.product` for ndarray and masked?